src/source/source_cache.js

```diff
@@ -27,23 +27,26 @@ class SourceCache extends Evented {
         this.id = id;
         this.dispatcher = dispatcher;
 
-        this.on('source.load', function() {
-            this._sourceLoaded = true;
-        });
-
-        this.on('error', function() {
-            this._sourceErrored = true;
-        });
-
-        this.on('data', function(event) {
-            if (this._sourceLoaded && event.dataType === 'source') {
+        this.on('data', function(e) {
+            // this._sourceLoaded signifies that the TileJSON is loaded if applicable
+            // if the source type does not come with a TileJSON, the flag signifies the
+            // source data has loaded (i.e geojson has been tiled on the worker and is ready)
+            if (e.dataType === 'source' && e.metadata) this._sourceLoaded = true;
+
+            // for sources with mutable data, this event fires when the underlying data
+            // to a source is changed. (i.e. GeoJSONSource#setData and ImageSource#serCoordinates)
+            if (this._sourceLoaded && e.dataType==="source" && e.update) {
```

Comment:
This is great ❤️ 

Notes/questions:

1) We should update the internal documentation [here](https://github.com/mapbox/mapbox-gl-js/blob/master/src/source/source.js#L53-L54) to reflect these changes to the Source interface contract.

2) Is `e.metadata` too generic a name?  I'm not even sure if I like this idea yet, but what about something like `e.sourceDataType == 'metadata'` and `e.sourceDataType == 'update'`?  (Basically having `sourceDataType` be analogous to `dataType`, but specific to source data events.)

3) Is there still a possibility of a race condition here for vector sources?  We need `SourceCache#update()` to be called after the TileJSON has loaded.  This currently happens either here (for events with `e.update`) or via `Style#_updateSources()`.   I'm imagining this problematic timeline:
- map.addSource('mysource', ...) (plus map.addLayer() using `mysource`) => requests TileJSON, marks `map._sourceDirty`.
- Next render frame fires; map._sourcesDirty being true leads to `style._updateSources()`... but since the TileJSON is still in flight, we don't get a `SourceCache#update()` for `mysource`.
- TileJSON lands => VectorTileSource fires `data` with `e.metadata` => `mysource`'s SourceCache marks itself `_sourceLoaded`.
- But now, the actual tiles for `mysource` are not loaded until something _else_ causes `_sourcesDirty` flag to be set.